subiquity/common/ubuntu_advantage.py

```diff
@@ -0,0 +1,151 @@
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+""" This module defines utilities to interface with Ubuntu Advantage
+subscriptions. """
+
+from abc import ABC, abstractmethod
+from datetime import datetime as dt
+import json
+import logging
+from subprocess import CalledProcessError, CompletedProcess
+import asyncio
+
+from subiquitycore import utils
+
+
+log = logging.getLogger("subiquitycore.common.ubuntu_advantage")
+
+
+class InvalidUATokenError(Exception):
+    """ Exception to be raised when the supplied token is invalid. """
+    def __init__(self, token: str, message: str = "") -> None:
+        self.token = token
+        self.message = message
+        super().__init__(message)
+
+
+class ExpiredUATokenError(Exception):
+    """ Exception to be raised when the supplied token has expired. """
+    def __init__(self, token: str, expires: str, message: str = "") -> None:
+        self.token = token
+        self.expires = expires
+        self.message = message
+        super().__init__(message)
+
+
+class CheckSubscriptionError(Exception):
+    """ Exception to be raised when we are unable to fetch information about
+    the Ubuntu Advantage subscription. """
+    def __init__(self, token: str, message: str = "") -> None:
+        self.token = token
+        self.message = message
+        super().__init__(message)
+
+
+class UAInterfaceStrategy(ABC):
+    """ Strategy to query information about a UA subscription. """
+    @abstractmethod
+    async def query_info(token: str) -> dict:
+        """ Return information about the UA subscription based on the token
+        provided.  """
+
+
+class MockedUAInterfaceStrategy(UAInterfaceStrategy):
+    """ Mocked version of the Ubuntu Advantage interface strategy. The info it
+    returns is based on example files and appearance of the UA token. """
+    def __init__(self, scale_factor: int = 1):
+        self.scale_factor = scale_factor
+        super().__init__()
+
+    async def query_info(self, token: str) -> dict:
+        """ Return the subscription info associated with the supplied
+        UA token. No actual query is done to the UA servers in this
+        implementation. Instead, we create a response based on the following
+        rules:
+        * Tokens starting with "x" will be considered expired.
+        * Tokens starting with "i" will be considered invalid.
+        * Tokens starting with "f" will generate an internal error.
+        """
+        await asyncio.sleep(1 / self.scale_factor)
```

Comment:
We want to leverage the environment variable for consistency with scripts/runtests.sh.  See also https://github.com/canonical/subiquity/blob/1aac580dfd6e70aee6ea1004bf64c03816591f83/subiquitycore/core.py#L69-L70

Reply:
Do you have a feel for how long the validation should normally take?  If significantly longer than 1 second, that would be nice to use before dividing by the timescale.

Reply:
Thanks, I'll make sure to leverage on `SUBIQUITY_REPLAY_TIMESCALE` when creating the `MockedUAInterfaceStrategy` here : https://github.com/canonical/subiquity/pull/1151/files/4fffca685e1047cec6ac620d9ad4a7f37527715f#diff-6363a69b28f40d7cbb460b917d2f53ac172366824645fc7e8a39b024b907a771R51

In normal conditions, the validation seems to be almost instantaneous. But it can take a while is there is some network misconfiguration.
The reason why I added a `sleep` here is so that the "Loading ..." screen has the opportunity to be visible in dry-mode for more than a few milliseconds. (so that we can test what happens it the user clicks on "cancel").

If we leave it to one second without using the timescale, it significantly slows down the unit tests - although we could also use some mocking to avoid the sleep.

What do you think is the best way?

Reply:
I think we're in agreement.